### DISCUSSION TITLE
Redundant definition splitting with Option.DEFINITIONS_FOR_ALL_OBJECTS

### DISCUSSION THREAD

**Discussion by roded** [victools/jsonschema-generator#337] [Q&A] [answered]
Hello,

I have a similar object model as described in https://github.com/victools/jsonschema-generator/issues/280 but where each subtype has additional properties which I'd like to be generated with definitions in the schema.

When specifying `Option.DEFINITIONS_FOR_ALL_OBJECTS`, the schema reverts to using "-1" and "-2" suffixes for the base and sub types (as described in issue https://github.com/victools/jsonschema-generator/issues/280).

Is there a way to get the desired effect of merging base and super types in the generated schema when `Option.DEFINITIONS_FOR_ALL_OBJECTS` is also specified (in addition to `JacksonOption.ALWAYS_REF_SUBTYPES` and `Option.DEFINITIONS_FOR_MEMBER_SUPERTYPES`)? 

Many thanks.

**Comment by CarstenWickner** [comment]
Hi @roded,

Can you perhaps provide an example code (representative type + your current configuration) and the desired (and potentially current) outcome?
This is otherwise quite abstract I'm afraid.

**Reply by roded** [reply]
Hi @CarstenWickner, 
Thanks for the reply.  Of course:

**Types**
```java
@JsonTypeInfo(use = JsonTypeInfo.Id.NAME)
@JsonSubTypes({
    @JsonSubTypes.Type(value = SubType1.class, name = "subtype1"),
    @JsonSubTypes.Type(value = SubType2.class, name = "subtype2"),
})
public interface BaseType {

}

@Value
public class SubType1 implements BaseType {

    NestedObject nestedObject;
}

@Value
public class SubType2 implements BaseType {

    List<BaseType> conditions;
}

@Value
public class NestedObject {

    Object field;
}
```

(`@Value` is a lombok annotations, these are simple POJOs).

**Config**

```java
        JacksonModule jacksonModule = new JacksonModule(JacksonOption.ALWAYS_REF_SUBTYPES);
        JavaxValidationModule javaxValidationModule =
            new GroupedAnnotationAwareJavaxValidationModule(JavaxValidationOption.INCLUDE_PATTERN_EXPRESSIONS);
        SchemaGeneratorConfigBuilder configBuilder =
            new SchemaGeneratorConfigBuilder(SchemaVersion.DRAFT_2020_12, OptionPreset.PLAIN_JSON).with(jacksonModule)
                .with(javaxValidationModule);
        configBuilder.with(Option.DEFINITIONS_FOR_MEMBER_SUPERTYPES);
        configBuilder.with(Option.DEFINITIONS_FOR_ALL_OBJECTS);
        configBuilder.with(Option.DEFINITION_FOR_MAIN_SCHEMA);
```

**Generated Schema**
```json
{
  "$schema" : "https://json-schema.org/draft/2020-12/schema",
  "$ref" : "#/$defs/BaseType",
  "$defs" : {
    "BaseType" : {
      "anyOf" : [ {
        "$ref" : "#/$defs/SubType1-2"
      }, {
        "$ref" : "#/$defs/SubType2-2"
      } ]
    },
    "NestedObject" : {
      "type" : "object",
      "properties" : {
        "field" : { }
      },
      "required" : [ "field" ]
    },
    "SubType1-1" : {
      "type" : "object",
      "properties" : {
        "nestedObject" : {
          "$ref" : "#/$defs/NestedObject"
        }
      },
      "required" : [ "nestedObject" ]
    },
    "SubType1-2" : {
      "$ref" : "#/$defs/SubType1-1",
      "type" : "object",
      "properties" : {
        "@type" : {
          "const" : "subtype1"
        }
      },
      "required" : [ "@type" ]
    },
    "SubType2-1" : {
      "type" : "object",
      "properties" : {
        "conditions" : {
          "type" : "array",
          "items" : {
            "$ref" : "#/$defs/BaseType"
          }
        }
      },
      "required" : [ "conditions" ]
    },
    "SubType2-2" : {
      "$ref" : "#/$defs/SubType2-1",
      "type" : "object",
      "properties" : {
        "@type" : {
          "const" : "subtype2"
        }
      },
      "required" : [ "@type" ]
    }
  }
}
```

I would like to avoid having the SubType classes be split into two definitions (-1 and -2).
Much obliged.

**Reply by CarstenWickner** [reply, marked as answer]
HI @roded,

I finally got around to finalizing the new `JacksonOption.INLINE_TRANSFORMED_SUBTYPES` for this.

I've added your example here:
https://github.com/victools/jsonschema-generator/blob/main/jsonschema-examples/src/main/java/com/github/victools/jsonschema/examples/JacksonSubtypeDefinitionExample.java

The output with the new `JacksonOption` looks like this:
https://github.com/victools/jsonschema-generator/blob/main/jsonschema-examples/src/test/resources/com/github/victools/jsonschema/examples/JacksonSubtypeDefinitionExample-result.json

It'll be available with the next release then (4.30.0). 😃

**Reply by roded** [reply]
@CarstenWickner wow! Amazing!
Thanks so much for the effort! I'll try once released.

**Reply by CarstenWickner** [reply]
Release [v4.30.0](https://github.com/victools/jsonschema-generator/releases/tag/v4.30.0) has been published. 🚀
